### ISSUE TITLE
Exception using `logging` within `gevent.threadpool.ThreadPool`

### ISSUE THREAD

**Issue by jaddison** [gevent/gevent#890]
* gevent version: 1.1.2
* Python version: 3.5.2
* Operating System: macos 10.12.1

**Edit**: Confirmed the same happens in Ubuntu 14.04 under Python 2.7.6. Changing the logging level to `WARNING` (which effectively disables the `logger.debug()` call in the thread function) allows everything to run smoothly.

### Description:

Trying to use python's `logging` within the thread of a `gevent.threadpool.ThreadPool` pool, running inside `gunicorn` running with its `gevent` worker class.

### What I've run:
The Python requirements:

```
falcon==1.1.0
gevent==1.1.2
gunicorn==19.6.0
```

My `app.py`:

```
import gevent.threadpool

import falcon
import logging

logging.basicConfig(level='DEBUG')
logger = logging.getLogger(__name__)

pool = gevent.threadpool.ThreadPool(4)

class GeventLogging(object):
    @classmethod
    def test(cls, n):
        logger.debug('test msg %s', n)

    def on_get(self, req, resp):
        pool.apply(self.test, args=(1,))
        resp.body = '{"message": "Hello world!"}'
        resp.status = falcon.HTTP_200

app = application = falcon.API()
app.add_route('/gevent', GeventLogging())
```

Run it with:

```
gunicorn -kgevent -w1 -b0.0.0.0:8001 app
```

Run `siege` against it like so:

```
siege -c10 -b "http://localhost:8001/gevent"
```

The exception I'm seeing occasionally:
 
```
[2016-11-09 09:43:43 -0800] [39685] [ERROR] Error handling request /gevent
Traceback (most recent call last):
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gunicorn/workers/async.py", line 52, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gunicorn/workers/ggevent.py", line 152, in handle_request
    super(GeventWorker, self).handle_request(*args)
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gunicorn/workers/async.py", line 103, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/falcon/api.py", line 209, in __call__
    responder(req, resp, **params)
  File "/Users/jaddison/projects/eemagery/gevent_threadpool_logging/app.py", line 17, in on_get
    pool.apply(self.test, args=(1,))
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gevent/pool.py", line 300, in apply
    return self.spawn(func, *args, **kwds).get()
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gevent/event.py", line 373, in get
    return self.get(block=False)
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gevent/event.py", line 363, in get
    return self._raise_exception()
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gevent/event.py", line 343, in _raise_exception
    reraise(*self.exc_info)
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gevent/hub.py", line 38, in reraise
    raise value.with_traceback(tb)
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gevent/threadpool.py", line 207, in _worker
    value = func(*args, **kwargs)
  File "/Users/jaddison/projects/eemagery/gevent_threadpool_logging/app.py", line 14, in test
    logger.debug('test msg %s', n)
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 1267, in debug
    self._log(DEBUG, msg, args, **kwargs)
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 1415, in _log
    self.handle(record)
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 1425, in handle
    self.callHandlers(record)
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 1487, in callHandlers
    hdlr.handle(record)
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 853, in handle
    self.acquire()
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/logging/__init__.py", line 804, in acquire
    self.lock.acquire()
  File "/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 145, in acquire
    rc = self._block.acquire(blocking, timeout)
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gevent/thread.py", line 74, in acquire
    return BoundedSemaphore.acquire(self, blocking, timeout)
  File "gevent/_semaphore.py", line 198, in gevent._semaphore.Semaphore.acquire (gevent/gevent._semaphore.c:4117)
  File "gevent/_semaphore.py", line 226, in gevent._semaphore.Semaphore.acquire (gevent/gevent._semaphore.c:3944)
  File "gevent/_semaphore.py", line 166, in gevent._semaphore.Semaphore._do_wait (gevent/gevent._semaphore.c:3178)
  File "/Users/jaddison/.virtualenvs/gevent_threadpool_logging/lib/python3.5/site-packages/gevent/hub.py", line 609, in switch
    return greenlet.switch(self)
gevent.hub.LoopExit: ('This operation would block forever', <Hub at 0x10992ce88 select pending=0 ref=0>)
```

**Comment by jamadden** [comment]
In a nutshell, you're using the gevent blocking API inside a thread with no other greenlets to switch to, and obviously that's not possible. (You only see it intermittently because the semaphore that the logger uses isn't always contended. This was [discussed recently](https://github.com/gevent/gevent/issues/863#issuecomment-249840775).) You would have the same problem using a socket in the threadpool or any other object that uses the gevent blocking API.

The general solution is to avoid using the blocking API inside a thread with no other greenlets (either don't use the API, or make sure there are other greenlets). (This specific problem with semaphores can be worked around by avoiding monkey-patching of `thread`, but I don't think that's possible under gunicorn.)

**Comment by jaddison** [comment]
> make sure there are other greenlets

How would I go about doing this - and what would be the consequences?

**Comment by jamadden** [comment]
You would need to spawn multiple greenlets in that single thread. What they would do, and thus the consequences, would be up to you.

**Comment by jaddison** [comment]
I won't lie, that response sounds vaguely ominous... is there something I should be aware of, spawning greenlets in `gevent.threadpool.ThreadPool` threads to avoid this `logging` issue?

**Comment by jaddison** [comment]
If I replace my `app.py` contents with the following, it works fine, but I have some questions.

```
import gevent.threadpool

import falcon
import logging

logging.basicConfig(level='DEBUG')
logger = logging.getLogger(__name__)

print(logging.Logger.manager.loggerDict)

pool = gevent.threadpool.ThreadPool(4)


class GeventLogging(object):
    @classmethod
    def test(cls, n):
        def bar():
            while True:
                gevent.sleep(0)
        t = gevent.spawn(bar)

        logger.debug('test msg %s', n)
        t.kill()

    def on_get(self, req, resp):
        pool.apply(self.test, args=(1,))
        resp.body = '{"message": "Hello world!"}'
        resp.status = falcon.HTTP_200

app = application = falcon.API()
app.add_route('/gevent', GeventLogging())
```

1. Do I need to call `t.kill()` above? or when the ThreadPool function (ie. `test()`) dies, will the green thread (ie. `bar()`) be automatically killed/terminated? I see no exceptions/errors, with or without `t.kill()`, so I'm curious.
2. is this the only way to accomplish what I'm looking to do: achieve logging in a ThreadPool thread function?

**Comment by jaddison** [comment]
Hmmm, while that works for logging, I don't think the `t.kill()` works as I'd expect. My `siege` session ended and the processes were pinning the CPU still, presumable in the `while True:` loop in `bar()`.

**Comment by jamadden** [comment]
If all you're really concerned about is logging from a threadpool thread, then your safest bet is to either restore or remove the lock that's causing the problem. 

```python
orig_rlock = gevent.monkey.get_original('threading', 'RLock')
for handler is threading._handlers.values():
   # Note that this causes logging to be non-cooperative; depending on the handler
   # that may or may not matter
   handler.lock = orig_rlock()
   # If you're daring and don't care about intermixed writes (again, depending on the handler
   # that may or may not matter) you could do this
   handler.lock = None
```

Note 1: gevent objects generally have thread-affinity and are not meant to be shared among threads, so, e.g., creating a gevent socket and passing it to a threadpool task will have similar issues; even this use of a gevent semaphore isn't *necessarily* safe, something I should have pointed out earlier (though in practice I think this does happen to be safe).

Note 2: I *think* this pattern works for general uses of the gevent blocking API in a threadpool; I don't tend to do that, though, so I've only played with it on a small example:

```python
def thread_task():
   def real_task():
      # use the gevent blocking API
  gevent.spawn(real_task)
  gevent.wait()

gevent.get_hub().threadpool.spawn(thread_task)
```

**Comment by jamadden** [comment]
Closing since there have been no updates.

**Comment by advance512** [comment]
This is probably fixed now, due to #1617.
